src/compiler/checker.ts

```diff
@@ -7771,6 +7771,11 @@ namespace ts {
                 if (target.flags & TypeFlags.Union && containsType(targetTypes, source)) {
                     return Ternary.True;
                 }
+                const discriminantType = findMatchingDiscriminantType(source, target);
+                if (discriminantType) {
+                    return isRelatedTo(source, discriminantType, reportErrors);

```

Comment:
I would run the entire loop with `false` for the `reportErrors` parameter. Then, if nothing matches, find the first constituent with a matching discriminant and call `isRelatedTo` with error reporting, or call `isRelatedTo` on the last one if no discriminants match.